# Experiment fd0fd28bb537dfce781410d0e384848efb166bc67495a71af2b8e88ad0971cef
# Base 09528c39a8284d40e07ffa68732a642c71025cdf
diff --git a/include/utmarray.h b/include/utmarray.h
index 2db30cbf6..44c4d545b 100644
--- a/include/utmarray.h
+++ b/include/utmarray.h
@@ -30,11 +30,11 @@ public:
         this->maximumSizeValue = 0;
     }
 
-    T& operator[](int index)
+    T& operator[](int v)
     {
-        if (index > this->maximumSizeValue - 1)
-            this->resize(index + 1);
-        return this->value[index];
+        if (v > this->maximumSizeValue - 1)
+            this->resize(v + 1);
+        return this->value[v];
     }
 
     int enlarge(int amount)
@@ -51,7 +51,7 @@ public:
         return this->desiredNumberValue;
     }
 
-    int contains(const T& item);
+    int contains(const T& r);
 
     int position(T& item)
     {
@@ -63,29 +63,28 @@ public:
         return -1;
     }
 
-    int add(T item);
-
-    int remove(T item)
+    int add(T r)
     {
-        int i = 0;
-        int maximum = this->maximumSizeValue;
-
-        for (; i < maximum; i++) {
-            if (this->value[i] == item)
-                break;
-        }
-
-        if (i >= maximum)
+        if (this->contains(r) == 1)
             return 0;
+        if (this->numberValue > this->maximumSizeValue - 1)
+            this->resize(this->numberValue + 1);
+        this->value[this->numberValue] = r;
+        this->numberValue++;
+        return 1;
+    }
 
-        maximum--;
-        if (i < maximum) {
-            do {
-                this->value[i] = this->value[i + 1];
-                i++;
-            } while (i < this->maximumSizeValue - 1);
-        }
-
+    int remove(T r)
+    {
+        int i;
+        for (i = 0;
+             i < this->maximumSizeValue && this->value[i] != r;
+             i++)
+            ;
+        if (i >= this->maximumSizeValue)
+            return 0;
+        for (; i < this->maximumSizeValue - 1; i++)
+            this->value[i] = this->value[i + 1];
         this->numberValue--;
         if (this->numberValue < 0)
             this->numberValue = 0;
@@ -134,29 +133,26 @@ public:
         }
     }
 
-public:
-    // PDB marks resize protected. Reconstructed callers still require public
-    // access, so keep the body here while the source-access boundary is open.
-    int resize(int new_size)
+protected:
+    int resize(int s)
     {
-        T* new_value = new T[new_size];
+        T* new_value = new T[s];
         if (new_value == 0)
             return 0;
 
         int i;
         for (i = 0;
-             i < this->maximumSizeValue && i < new_size;
+             i < this->maximumSizeValue && i < s;
              i++) {
             new_value[i] = this->value[i];
         }
 
         delete[] this->value;
         this->value = new_value;
-        this->maximumSizeValue = new_size;
+        this->maximumSizeValue = s;
         return 1;
     }
 
-public:
     T* value;
     int numberValue;
     int desiredNumberValue;
@@ -164,52 +160,16 @@ public:
 };
 
 template <class T>
-int ManagedArray<T>::contains(const T& item)
+int ManagedArray<T>::contains(const T& r)
 {
     int i;
-    T wanted;
-    T current;
-
     for (i = 0; i < this->numberValue; i++) {
         if (i >= this->maximumSizeValue)
             return 0;
-        current = this->value[i];
-        wanted = item;
-        if (current == wanted)
+        if (this->value[i] == r)
             return 1;
     }
     return 0;
 }
 
-template <class T>
-int ManagedArray<T>::add(T val)
-{
-    int i;
-
-    for (i = 0; i < this->numberValue; i++) {
-        if (i >= this->maximumSizeValue)
-            break;
-        if (this->value[i] == val)
-            return 0;
-    }
-    if (this->numberValue > this->maximumSizeValue - 1) {
-        int newSize = this->numberValue + 1;
-        T* newValue = (T*)::operator new(newSize * sizeof(T));
-        if (newValue != 0) {
-            int j;
-            for (j = 0; j < this->maximumSizeValue; j++) {
-                if (j >= newSize)
-                    break;
-                newValue[j] = this->value[j];
-            }
-            ::operator delete(this->value);
-            this->value = newValue;
-            this->maximumSizeValue = newSize;
-        }
-    }
-    this->value[this->numberValue] = val;
-    this->numberValue++;
-    return 1;
-}
-
 #endif
diff --git a/src/stat_obj.cpp b/src/stat_obj.cpp
index 445a88f9a..a82694d01 100644
--- a/src/stat_obj.cpp
+++ b/src/stat_obj.cpp
@@ -1339,19 +1339,14 @@ uchar RGE_Static_Object::update()
         this->player_object_node = this->owner->addObject(this, 1, this->dopple_flag);
         return 0;
     }
-    if (this->groupMembers.numberValue > 0) {
+    if (this->groupMembers.number() > 0) {
         int i = 0;
-        if (i < this->groupMembers.numberValue) {
-            while (i < this->groupMembers.numberValue) {
+        if (i < this->groupMembers.number()) {
+            while (i < this->groupMembers.number()) {
                 int id = this->groupMembers[i];
                 RGE_Static_Object* obj = this->owner->world->object(id);
                 if (obj == 0 || obj->object_state > 2) {
-                    int n = this->groupMembers.maximumSizeValue; int k = 0;
-                    while (k < n && this->groupMembers.value[k] != id) { k++; }
-                    if (k < n) {
-                        if (k < n - 1) { do { int* p = this->groupMembers.value; k++; p[k-1] = p[k]; } while (k < this->groupMembers.maximumSizeValue - 1); }
-                        int gn = this->groupMembers.numberValue - 1; this->groupMembers.numberValue = gn; if (gn < 0) this->groupMembers.numberValue = 0;
-                    }
+                    this->groupMembers.remove(id);
                     i--;
                 }
                 i++;
@@ -1624,23 +1619,23 @@ void RGE_Static_Object::save(int file)
     rge_write(saved_file, &this->groupCommanderValue, 4);
     rge_write(saved_file, &this->groupRangeValue, 4);
 
-    file = this->groupMembers.numberValue;
+    file = this->groupMembers.number();
     rge_write(saved_file, &file, 4);
     i = 0;
-    if (this->groupMembers.numberValue > 0) do {
+    if (this->groupMembers.number() > 0) do {
         file = this->groupMembers[i];
         rge_write(saved_file, &file, 4);
         ++i;
-    } while (i < this->groupMembers.numberValue);
+    } while (i < this->groupMembers.number());
 
-    file = this->pathingGroupMembers.numberValue;
+    file = this->pathingGroupMembers.number();
     rge_write(saved_file, &file, 4);
     int k = 0;
-    if (this->pathingGroupMembers.numberValue > 0) do {
+    if (this->pathingGroupMembers.number() > 0) do {
         file = this->pathingGroupMembers[k];
         rge_write(saved_file, &file, 4);
         ++k;
-    } while (k < this->pathingGroupMembers.numberValue);
+    } while (k < this->pathingGroupMembers.number());
 
     this->sprite_list->save(saved_file);
 }
@@ -2223,10 +2218,10 @@ void RGE_Static_Object::set_object_state(unsigned char state)
         if (state == 3 &&
             this->groupCommanderValue == this->id &&
             this->unitAIValue != 0 &&
-            this->groupMembers.numberValue > 1) {
+            this->groupMembers.number() > 1) {
             int i = 0;
             int newCommander = -1;
-            if (this->groupMembers.numberValue > 0) {
+            if (this->groupMembers.number() > 0) {
                 do {
                     if (this->groupMembers[i] != this->id) {
                         RGE_Static_Object* object =
@@ -2244,7 +2239,7 @@ void RGE_Static_Object::set_object_state(unsigned char state)
                         }
                     }
                     ++i;
-                } while (i < this->groupMembers.numberValue);
+                } while (i < this->groupMembers.number());
             }
         }
 
@@ -3452,28 +3447,12 @@ int RGE_Static_Object::isGroupCommander(void) {
 
 
 
-int RGE_Static_Object::inGroup() { return this->groupMembers.numberValue > 0; }
+int RGE_Static_Object::inGroup() { return this->groupMembers.number() > 0; }
 
 
 
 int RGE_Static_Object::unitIsInGroup(int value) {
-    int count = this->groupMembers.numberValue;
-    int index = 0;
-
-    if (count > 0) {
-        for (; index < count; index++) {
-            if (index >= this->groupMembers.maximumSizeValue) {
-                goto not_found_shared;
-            }
-            if (this->groupMembers.value[index] == value) {
-                return 1;
-            }
-        }
-    }
-    return 0;
-
-not_found_shared:
-    return 0;
+    return this->groupMembers.contains(value);
 }
 
 
@@ -3485,7 +3464,7 @@ int RGE_Static_Object::createGroup(int* list, int count, int commander, float ra
         return 0;
     }
 
-    groupMembers.numberValue = 0;
+    groupMembers.zeroNumber();
 
     if (count > 0) {
         int* current = list;
@@ -3493,37 +3472,9 @@ int RGE_Static_Object::createGroup(int* list, int count, int commander, float ra
         do {
             int id = *current;
             if (id >= 0) {
-                int number = groupMembers.numberValue;
-                int i = 0;
-                while (i < number) {
-                    if (i >= groupMembers.maximumSizeValue) {
-                        break;
-                    }
-                    if (groupMembers.value[i] == id) {
-                        goto next;
-                    }
-                    i++;
-                }
-
-                if (number > groupMembers.maximumSizeValue - 1) {
-                    int newMaximum = number + 1;
-                    int* newValue = new int[newMaximum];
-                    if (newValue != 0) {
-                        i = 0;
-                        for (; i < groupMembers.maximumSizeValue && i < newMaximum; i++) {
-                            newValue[i] = groupMembers.value[i];
-                        }
-                        delete[] groupMembers.value;
-                        groupMembers.value = newValue;
-                        groupMembers.maximumSizeValue = newMaximum;
-                    }
-                }
-
-                groupMembers.value[groupMembers.numberValue] = id;
-                groupMembers.numberValue = groupMembers.numberValue + 1;
+                groupMembers.add(id);
             }
 
-        next:
             current = current + 1;
             remaining = remaining - 1;
         } while (remaining != 0);
@@ -3564,42 +3515,11 @@ int RGE_Static_Object::commanderCreateGroup(int* list, int count, int param3, fl
 
 int RGE_Static_Object::addToGroup(int groupIndex, float unused)
 {
-    if (this->groupMembers.numberValue >= 25) {
+    if (this->groupMembers.number() >= 25) {
         return 0;
     }
 
-    {
-        ManagedArray<int>& members = this->groupMembers;
-        int count = members.numberValue;
-        int i = 0;
-        while (i < count) {
-            if (i >= members.maximumSizeValue) {
-                break;
-            }
-            if (members.value[i] == groupIndex) {
-                goto done;
-            }
-            i++;
-        }
-
-        if (count > members.maximumSizeValue - 1) {
-            int newMax = count + 1;
-            int* newArray = new int[newMax];
-            if (newArray != 0) {
-                for (i = 0; i < members.maximumSizeValue && i < newMax; i++) {
-                    newArray[i] = members.value[i];
-                }
-                delete[] members.value;
-                members.value = newArray;
-                members.maximumSizeValue = newMax;
-            }
-        }
-
-        members.value[members.numberValue] = groupIndex;
-        members.numberValue++;
-    }
-
-done:
+    this->groupMembers.add(groupIndex);
     return 1;
 }
 
@@ -3615,12 +3535,12 @@ int RGE_Static_Object::commanderAddToGroup(int id, float range)
         return 0;
     }
 
-    if (group.numberValue >= 25) {
+    if (group.number() >= 25) {
         return 0;
     }
 
     int i = 0;
-    if (group.numberValue > 0) {
+    if (group.number() > 0) {
         do {
             int member = group[i];
             temp[i] = member;
@@ -3632,7 +3552,7 @@ int RGE_Static_Object::commanderAddToGroup(int id, float range)
                 }
             }
             ++i;
-        } while (i < group.numberValue);
+        } while (i < group.number());
     }
 
     temp[i] = id;
@@ -3646,40 +3566,8 @@ int RGE_Static_Object::commanderAddToGroup(int id, float range)
 
 
 int RGE_Static_Object::removeFromGroup(int param_1) {
-    int count = groupMembers.maximumSizeValue;
-    int index = 0;
-    int result;
-
-    if (count > 0) {
-        int* ptr = groupMembers.value;
-        while (1) {
-            if (*ptr == param_1) break;
-            ++index;
-            ++ptr;
-            if (index < count) continue;
-            break;
-        }
-    }
-    if (index >= count) {
-        result = 0;
-        goto final;
-    }
-
-    if (index < groupMembers.maximumSizeValue - 1) {
-        while (index < groupMembers.maximumSizeValue - 1) {
-            groupMembers.value[index] = groupMembers.value[index + 1];
-            ++index;
-        }
-    }
-
-    --groupMembers.numberValue;
-    if (groupMembers.numberValue < 0) {
-        groupMembers.numberValue = 0;
-    }
-    result = 1;
-
-final:
-    if (groupMembers.numberValue <= 0) {
+    int result = groupMembers.remove(param_1);
+    if (groupMembers.number() <= 0) {
         groupCommanderValue = -1;
     }
     return result;
@@ -3695,66 +3583,15 @@ int RGE_Static_Object::commanderRemoveFromGroup(int id)
     }
 
     int i = 0;
-    if (this->groupMembers.numberValue > 0) {
+    if (this->groupMembers.number() > 0) {
         do {
-            if (i > this->groupMembers.maximumSizeValue - 1) {
-                int newMax = i + 1;
-                int* newValue = new int[newMax];
-                if (newValue != 0) {
-                    int j = 0;
-                    if (this->groupMembers.maximumSizeValue > 0) {
-                        do {
-                            if (j < newMax) {
-                                ++j;
-                                newValue[j - 1] = this->groupMembers.value[j - 1];
-                            } else {
-                                break;
-                            }
-                        } while (j < this->groupMembers.maximumSizeValue);
-                    }
-                    delete[] this->groupMembers.value;
-                    this->groupMembers.value = newValue;
-                    this->groupMembers.maximumSizeValue = newMax;
-                }
-            }
-
-            RGE_Static_Object* object = this->owner->world->object(this->groupMembers.value[i]);
+            RGE_Static_Object* object = this->owner->world->object(this->groupMembers[i]);
             if (object != 0 && object->id != this->id) {
-                int max = object->groupMembers.maximumSizeValue;
-                int j = 0;
-                if (max > 0) {
-                    int* p = object->groupMembers.value;
-                    while (1) {
-                        if (*p == id) {
-                            break;
-                        }
-                        ++j;
-                        ++p;
-                        if (j < max) {
-                            continue;
-                        }
-                        break;
-                    }
-                }
-
-                if (j < max) {
-                    if (j < max - 1) {
-                        do {
-                            object->groupMembers.value[j] = object->groupMembers.value[j + 1];
-                            ++j;
-                        } while (j < object->groupMembers.maximumSizeValue - 1);
-                    }
-
-                    int newCount = object->groupMembers.numberValue - 1;
-                    object->groupMembers.numberValue = newCount;
-                    if (newCount < 0) {
-                        object->groupMembers.numberValue = 0;
-                    }
-                }
+                object->groupMembers.remove(id);
             }
 
             ++i;
-        } while (i < this->groupMembers.numberValue);
+        } while (i < this->groupMembers.number());
     }
 
     this->removeFromGroup(id);
@@ -3764,7 +3601,7 @@ int RGE_Static_Object::commanderRemoveFromGroup(int id)
 
 
 int RGE_Static_Object::destroyGroup() {
-    groupMembers.numberValue = 0;
+    groupMembers.zeroNumber();
     groupCommanderValue = -1;
     return 1;
 }
@@ -3773,7 +3610,7 @@ int RGE_Static_Object::destroyGroup() {
 int RGE_Static_Object::commanderDestroyGroup()
 {
     int i = 0;
-    if (this->groupMembers.numberValue > 0) {
+    if (this->groupMembers.number() > 0) {
         do {
             RGE_Static_Object* object =
                 this->owner->world->object(this->groupMembers[i]);
@@ -3781,7 +3618,7 @@ int RGE_Static_Object::commanderDestroyGroup()
                 object->destroyGroup();
             }
             ++i;
-        } while (i < this->groupMembers.numberValue);
+        } while (i < this->groupMembers.number());
     }
     this->destroyGroup();
     return 1;
@@ -3792,44 +3629,7 @@ int RGE_Static_Object::commanderDestroyGroup()
 
 void RGE_Static_Object::addToPathingGroup(int groupIndex)
 {
-    int i = 0;
-    int count = this->pathingGroupMembers.numberValue;
-    if (count > 0) {
-        int max = this->pathingGroupMembers.maximumSizeValue;
-        while (1) {
-            if (i >= max)
-                break;
-            if (this->pathingGroupMembers.value[i] == groupIndex)
-                return;
-            i++;
-            if (i < count)
-                continue;
-            break;
-        }
-    }
-
-    if (count > this->pathingGroupMembers.maximumSizeValue - 1) {
-        int newMax = count + 1;
-        int* newMembers = new int[newMax];
-        if (newMembers != 0) {
-            int max = this->pathingGroupMembers.maximumSizeValue;
-            int j = 0;
-            if (max > 0) {
-                for (; j < max; j++) {
-                    if (j >= newMax)
-                        break;
-                    newMembers[j] = this->pathingGroupMembers.value[j];
-                    max = this->pathingGroupMembers.maximumSizeValue;
-                }
-            }
-            delete[] this->pathingGroupMembers.value;
-            this->pathingGroupMembers.value = newMembers;
-            this->pathingGroupMembers.maximumSizeValue = newMax;
-        }
-    }
-
-    this->pathingGroupMembers.value[this->pathingGroupMembers.numberValue] = groupIndex;
-    this->pathingGroupMembers.numberValue++;
+    this->pathingGroupMembers.add(groupIndex);
 }
 
 
@@ -3837,7 +3637,7 @@ void RGE_Static_Object::addToPathingGroup(int groupIndex)
 
 void RGE_Static_Object::removeAllFromPathingGroup(void) {
     if (owner->computerPlayer() == 1) {
-        pathingGroupMembers.numberValue = 0;
+        pathingGroupMembers.zeroNumber();
     }
 }
 
